File: lib/core/diagnostics/diagnostic_log.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:path_provider/path_provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

class DiagnosticLog {
  DiagnosticLog._();

  static const _enabledKey = 'diagnostic.log.enabled';
  static const _sessionKey = 'diagnostic.log.session';
  static SharedPreferences? _prefs;
  static File? _file;
  static bool _enabled = false;
  static String _session = '';
  static Future<void> _pending = Future.value();

  static bool get enabled => _enabled;
  static String get session => _session;

  static Future<void> init(SharedPreferences prefs) async {
    _prefs = prefs;
    _enabled = prefs.getBool(_enabledKey) ?? false;
    _session = prefs.getString(_sessionKey) ?? '';
    if (_enabled && _session.isEmpty) {
      _session = _newSessionId();
      await prefs.setString(_sessionKey, _session);
    }
    final dir = await getApplicationDocumentsDirectory();
    _file = File('${dir.path}/tonari-diagnostic.jsonl');
  }

  static Future<void> startSession() async {
    _session = _newSessionId();
    _enabled = true;
    await _prefs!.setString(_sessionKey, _session);
    await _prefs!.setBool(_enabledKey, true);
    await clear();
    write('diagnostic', 'session_start', {'sessionId': _session});
  }

  static Future<void> stopSession() async {
    write('diagnostic', 'session_stop', {'sessionId': _session});
    await _pending;
    _enabled = false;
    await _prefs!.setBool(_enabledKey, false);
  }

  static void write(
    String source,
    String event, [
    Map<String, Object?> fields = const {},
  ]) {
    if (!_enabled) return;
    final file = _file!;
    final record = <String, Object?>{
      'ts': DateTime.now().toIso8601String(),
      'session': _session,
      'source': source,
      'event': event,
      ...fields,
    };
    final line = '${jsonEncode(record)}\n';
    _pending = _pending
        .then<void>((_) async {
          await file.writeAsString(line, mode: FileMode.append);
        })
        .catchError((Object _) {});
  }

  static Future<String> read() async {
    await _pending;
    final file = _file!;
    if (!await file.exists()) return '';
    return file.readAsString();
  }

  static Future<void> clear() async {
    await _pending;
    final file = _file!;
    await file.writeAsString('');
  }

  /// Writes the current log to a `.txt` in the temp dir for sharing, returning
  /// its path (or null when empty). Sharing via the system sheet needs no
  /// entitlement, so it works on any signing.
  static Future<String?> exportTxt() async {
    await _pending;
    final content = await read();
    if (content.isEmpty) return null;
    final dir = await getTemporaryDirectory();
    final name = _session.isEmpty ? 'tonari-diagnostic' : 'tonari-$_session';
    final out = File('${dir.path}/$name.txt');
    await out.writeAsString(content);
    return out.path;
  }

  static String _newSessionId() {
    return DateTime.now().microsecondsSinceEpoch.toRadixString(36);
  }
}

File: native/TonariCore/Sources/TonariCore/Diagnostics/DiagnosticLog.swift
import Foundation

/// Opt-in JSONL event log the user exports to report problems. Writes are
/// fire-and-forget and serialized on a private queue, so it can be called
/// from any thread, including media-engine callbacks.
public final class DiagnosticLog: @unchecked Sendable {
    public static let shared = DiagnosticLog(
        directory: URL.documentsDirectory,
        defaults: .standard
    )

    private static let enabledKey = "diagnostic.log.enabled"
    private static let sessionKey = "diagnostic.log.session"

    private let file: URL
    private let defaults: UserDefaults
    private let queue = DispatchQueue(label: "tonari.diagnostic-log")
    private var _enabled: Bool
    private var _session: String

    public init(directory: URL, defaults: UserDefaults) {
        file = directory.appending(path: "tonari-diagnostic.jsonl")
        self.defaults = defaults
        _enabled = defaults.bool(forKey: Self.enabledKey)
        _session = defaults.string(forKey: Self.sessionKey) ?? ""
        if _enabled && _session.isEmpty {
            _session = Self.newSessionId()
            defaults.set(_session, forKey: Self.sessionKey)
        }
    }

    public var enabled: Bool { queue.sync { _enabled } }
    public var session: String { queue.sync { _session } }

    public func startSession() {
        let session = Self.newSessionId()
        queue.sync {
            _session = session
            _enabled = true
            try? Data().write(to: file)
        }
        defaults.set(session, forKey: Self.sessionKey)
        defaults.set(true, forKey: Self.enabledKey)
        write("diagnostic", "session_start", ["sessionId": session])
    }

    /// Re-enables collection under the previous session id without clearing
    /// the log, so a stopped session can keep accumulating.
    public func resumeSession() {
        guard !session.isEmpty else { return startSession() }
        queue.sync { _enabled = true }
        defaults.set(true, forKey: Self.enabledKey)
        write("diagnostic", "session_resume", ["sessionId": session])
    }

    public func stopSession() {
        write("diagnostic", "session_stop", ["sessionId": session])
        queue.sync { _enabled = false }
        defaults.set(false, forKey: Self.enabledKey)
    }

    public func write(_ source: String, _ event: String, _ fields: [String: any Sendable] = [:]) {
        let ts = Date().formatted(Self.timestampStyle)
        queue.async { [self] in
            guard _enabled else { return }
            let head: [String: Any] = ["ts": ts, "session": _session, "source": source, "event": event]
            appendLine(Self.json(head, then: fields))
        }
    }

    public func read() -> String {
        queue.sync { (try? String(contentsOf: file, encoding: .utf8)) ?? "" }
    }

    public func clear() {
        queue.sync { try? Data().write(to: file) }
    }

    /// Copies the log to a `.txt` in the temp directory for the share sheet;
    /// nil when there is nothing to share.
    public func exportTxt() throws -> URL? {
        let content = read()
        guard !content.isEmpty else { return nil }
        let name = session.isEmpty ? "tonari-diagnostic" : "tonari-\(session)"
        let out = URL.temporaryDirectory.appending(path: "\(name).txt")
        try content.write(to: out, atomically: true, encoding: .utf8)
        return out
    }

    private func appendLine(_ line: String) {
        let data = Data((line + "\n").utf8)
        guard let handle = try? FileHandle(forWritingTo: file) else {
            try? data.write(to: file)
            return
        }
        handle.seekToEndOfFile()
        handle.write(data)
        try? handle.close()
    }

    /// Keeps ts/session/source/event first so lines stay readable, followed by
    /// the caller's fields.
    private static func json(_ head: [String: Any], then fields: [String: any Sendable]) -> String {
        func encode(_ object: [String: Any], keys: [String]) -> [String] {
            keys.map { key in
                let pair = try! JSONSerialization.data(withJSONObject: [key: object[key]!], options: [.withoutEscapingSlashes, .fragmentsAllowed])
                return String(decoding: pair.dropFirst().dropLast(), as: UTF8.self)
            }
        }
        let parts = encode(head, keys: ["ts", "session", "source", "event"])
            + encode(fields, keys: fields.keys.sorted())
        return "{" + parts.joined(separator: ",") + "}"
    }

    private static let timestampStyle = Date.ISO8601FormatStyle(includingFractionalSeconds: true, timeZone: .current)

    private static func newSessionId() -> String {
        String(Int64(Date().timeIntervalSince1970 * 1_000_000), radix: 36)
    }
}
